lib/web3-contracts.js

```diff
@@ -131,248 +106,189 @@ export function useTokenBalance(symbol) {
+    if ((!account && !address) || !tokenContract) {
       return
     }
 
     const onTransfer = (from, to, value) => {
-      if (from === account || to === account) {
+      if (
+        from === account ||
+        to === account ||
+        from === address ||
+        to === address
+      ) {
         updateBalance()
       }
     }
     tokenContract.on('Transfer', onTransfer)
 
-    // Filter transfers from and to the account
-    // const filters = [
-    //   tokenContract.filters.Transfer(account),
-    //   tokenContract.filters.Transfer(null, account),
-    // ]
-    // filters.forEach(filter => tokenContract.on(filter, onTransfer))
-
     return () => {
       tokenContract.removeListener('Transfer', onTransfer)
-
-      // filters.forEach(filter =>
-      //   tokenContract.removeListener(filter, onTransfer)
-      // )
     }
-  }, [account, tokenContract, updateBalance])
+  }, [account, address, tokenContract, updateBalance])
 
   return balance
 }
 
-export function useJurorRegistryAnjBalance() {
-  const { account } = useWeb3Connect()
-  const [balance, setBalance] = useState(bigNum(-1))
+export function useBondingCurvePrice(amount, forwards = true) {
+  const [loading, setLoading] = useState(false)
+  const [price, setPrice] = useState(bigNum(-1))
 
-  const tokenContract = useKnownContract('TOKEN_ANJ')
-  const jurorsRegistryContract = useKnownContract('JURORS_REGISTRY')
-  const wrapperContract = useKnownContract('WRAPPER')
+  const bancorContract = useKnownContract('BANCOR_FORMULA')
 
-  const cancelBalanceUpdate = useRef(null)
+  const [treasuryAddress] = getKnownContract('BONDING_CURVE_TREASURY')
+  const antTreasuryBalance = useTokenBalance('ANT', treasuryAddress)
 
-  const updateBalance = useCallback(() => {
-    let cancelled = false
-
-    if (cancelBalanceUpdate.current) {
-      cancelBalanceUpdate.current()
-      cancelBalanceUpdate.current = null
-    }
-
-    if (!account || !tokenContract || !jurorsRegistryContract) {
-      setBalance(bigNum(-1))
-      return
-    }
-
-    cancelBalanceUpdate.current = () => {
-      cancelled = true
-    }
-
-    jurorsRegistryContract.balanceOf(account).then(([activeBalance]) => {
-      if (!cancelled) {
-        setBalance(activeBalance)
-      }
-    })
-  }, [account, tokenContract, jurorsRegistryContract])
+  const { loading: supplyLoading, totalSupply } = useAnjTotalSupply()
 
   useEffect(() => {
-    // Always update the balance if updateBalance() has changed
-    updateBalance()
-
-    if (!wrapperContract || !account) {
+    let cancelled = false
+    let retryTimer
+
+    if (
+      antTreasuryBalance.eq(-1) ||
+      !bancorContract ||
+      supplyLoading ||
+      totalSupply.eq(-1)
+    ) {
       return
     }
-
-    const onBought = (from, to, value) => {
-      if (from === account) {
-        updateBalance()
+    const getSalePrice = async () => {
+      try {
+        setLoading(true)
+        const salePrice = await (forwards
```

Comment:
We may want a more descriptive name than `forwards`; it is going to be quite hard to tell what we mean in a few months.

We could extract some of these operations that change based on the direction into a utility file and use symbols or strings to denote what we mean (e.g. `OBTAIN_ANJ` and `RETURN_ANJ`)

Reply:
Will do this as we do the UI work; seems a better strategy than just abstracting this now.